### PR TITLE
[yargs] Fix parse() doesn't recognize (argv, callback)

### PR DESCRIPTION
**Pull request by xymopen**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [ ] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ ] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes:
  - [yargs API Reference](https://yargs.js.org/docs/#api-reference-parseargs-context-parsecallback)
  - [parse() in yargs@v8.0.2](https://github.com/yargs/yargs/blob/v8.0.2/yargs.js#L508)
  - [parse() in yargs@v10.1.2](https://github.com/yargs/yargs/blob/v10.1.2/yargs.js#L517)
  - [parse() in yargs@v11.1.0](https://github.com/yargs/yargs/blob/v11.1.0/yargs.js#L526)
  - [parse() in yargs@v12.0.5](https://github.com/yargs/yargs/blob/v12.0.5/yargs.js#L539)
  - [parse() in yargs@v13.3.0](https://github.com/yargs/yargs/blob/v13.3.0/yargs.js#L539)
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

**Comment by xymopen** [comment]
`npm test` shows:

```
Error: .vscode: filesystem doesn't support paths of the form './x'.
```

and `npm run lint yargs` shows: 

```
The 'no-redundant-jsdoc-2' rule threw an error in 'types/yargs/index.d.ts':
Error: Unexpected tag kind: JSDocDeprecatedTag
```

File: types/yargs/v10/index.d.ts [hunk 1]
```diff
@@ -152,8 +152,10 @@ declare namespace yargs {
         options(key: string, options: Options): Argv;
         options(options: { [key: string]: Options }): Argv;
 
-        parse(): Arguments;
-        parse(arg: string | string[], context?: object, parseCallback?: ParseCallback): Arguments;
+        parse(arg?: string | ReadonlyArray<string>): Arguments;
+        parse(arg: string | ReadonlyArray<string>, parseCallback: ParseCallback): Arguments;
+        parse(arg: string | ReadonlyArray<string>, context: object): Arguments;
+        parse(arg: string | ReadonlyArray<string>, context: object, parseCallback: ParseCallback): Arguments;
```
Comment (forivall) [inline types/yargs/v10/index.d.ts:158]: These two last signatures can be combined to
```ts
        parse(arg: string | ReadonlyArray<string>, context: object, parseCallback?: ParseCallback): Arguments;
```

File: types/yargs/v11/index.d.ts [hunk 2]
```diff
@@ -154,8 +154,10 @@ declare namespace yargs {
         options(key: string, options: Options): Argv;
         options(options: { [key: string]: Options }): Argv;
 
-        parse(): Arguments;
-        parse(arg: string | string[], context?: object, parseCallback?: ParseCallback): Arguments;
+        parse(arg?: string | ReadonlyArray<string>): Arguments;
+        parse(arg: string | ReadonlyArray<string>, parseCallback: ParseCallback): Arguments;
+        parse(arg: string | ReadonlyArray<string>, context: object): Arguments;
+        parse(arg: string | ReadonlyArray<string>, context: object, parseCallback: ParseCallback): Arguments;
```
Comment (forivall) [inline types/yargs/v11/index.d.ts:160]: ditto https://github.com/DefinitelyTyped/DefinitelyTyped/pull/46251/files#r459641607

File: types/yargs/v12/index.d.ts [hunk 3]
```diff
@@ -185,8 +185,10 @@ declare namespace yargs {
         options<K extends string, O extends Options>(key: K, options: O): Argv<T & { [key in K]: InferredOptionType<O> }>;
         options<O extends { [key: string]: Options }>(options: O): Argv<Omit<T, keyof O> & InferredOptionTypes<O>>;
 
-        parse(): { [key in keyof Arguments<T>]: Arguments<T>[key] };
-        parse(arg: string | ReadonlyArray<string>, context?: object, parseCallback?: ParseCallback<T>): { [key in keyof Arguments<T>]: Arguments<T>[key] };
+        parse(arg?: string | ReadonlyArray<string>): { [key in keyof Arguments<T>]: Arguments<T>[key] };
+        parse(arg: string | ReadonlyArray<string>, parseCallback: ParseCallback<T>): { [key in keyof Arguments<T>]: Arguments<T>[key] };
+        parse(arg: string | ReadonlyArray<string>, context: object): { [key in keyof Arguments<T>]: Arguments<T>[key] };
+        parse(arg: string | ReadonlyArray<string>, context: object, parseCallback: ParseCallback<T>): { [key in keyof Arguments<T>]: Arguments<T>[key] };
```
Comment (forivall) [inline types/yargs/v12/index.d.ts:191]: ditto https://github.com/DefinitelyTyped/DefinitelyTyped/pull/46251/files#r459641607

File: types/yargs/v13/index.d.ts [hunk 4]
```diff
@@ -415,8 +415,10 @@ declare namespace yargs {
          * Note: Providing a callback to parse() disables the `exitProcess` setting until after the callback is invoked.
          * @param [context]  Provides a useful mechanism for passing state information to commands
          */
-        parse(): { [key in keyof Arguments<T>]: Arguments<T>[key] };
-        parse(arg: string | ReadonlyArray<string>, context?: object, parseCallback?: ParseCallback<T>): { [key in keyof Arguments<T>]: Arguments<T>[key] };
+        parse(arg?: string | ReadonlyArray<string>): { [key in keyof Arguments<T>]: Arguments<T>[key] };
+        parse(arg: string | ReadonlyArray<string>, parseCallback: ParseCallback<T>): { [key in keyof Arguments<T>]: Arguments<T>[key] };
+        parse(arg: string | ReadonlyArray<string>, context: object): { [key in keyof Arguments<T>]: Arguments<T>[key] };
+        parse(arg: string | ReadonlyArray<string>, context: object, parseCallback: ParseCallback<T>): { [key in keyof Arguments<T>]: Arguments<T>[key] };
```
Comment (forivall) [inline types/yargs/v13/index.d.ts:421]: ditto https://github.com/DefinitelyTyped/DefinitelyTyped/pull/46251/files#r459641607

File: types/yargs/v8/index.d.ts [hunk 5]
```diff
@@ -14,7 +14,10 @@ declare namespace yargs {
     interface Argv {
         argv: Arguments;
         (args?: string[], cwd?: string): Arguments;
-        parse(args: string | string[], context?: object, parseCallback?: ParseCallback): Arguments;
+        parse(arg: string | ReadonlyArray<string>): Arguments;
+        parse(arg: string | ReadonlyArray<string>, parseCallback: ParseCallback): Arguments;
+        parse(arg: string | ReadonlyArray<string>, context: object): Arguments;
+        parse(arg: string | ReadonlyArray<string>, context: object, parseCallback: ParseCallback): Arguments;
```
Comment (forivall) [inline types/yargs/v8/index.d.ts:20]: ditto https://github.com/DefinitelyTyped/DefinitelyTyped/pull/46251/files#r459641607

File: types/yargs/v10/index.d.ts (hunk 1 above)
Comment (xymopen) [reply types/yargs/v10/index.d.ts:158]: In my case the combination makes the `parseCallback()` argument fit into the `context` parameter for its `object` type, and `parseCallback` parameter become `undefined`, resulting in the type of parameters for `parseCallback()` not get inferred.

**Comment by xymopen** [comment]
@forivall Would you please tell me how can I fix the keep-poping `no-redundant-jsdoc-2  JSDoc tag '@deprecated' is redundant in TypeScript code.` error in the CI pipeline?

**Review by forivall — APPROVED** [review]

**Comment by forivall** [comment]
Sorry, i don't know. I'm just a maintainer for yargs, you can try to reach out to the  @DefinitelyTyped/maintainers via [their gitter](https://gitter.im/DefinitelyTyped/DefinitelyTyped) as described in https://github.com/DefinitelyTyped/DefinitelyTyped#my-pr-is-merged-when-will-the-types-npm-package-be-updated

**Comment by forivall** [comment]
I did a bit of searching, and it looks like we just need to wait for a new version of dtslint to drop https://github.com/microsoft/dtslint/issues/299

**Comment by forivall** [comment]
@typescript-bot please re-run the build!

**Comment by xymopen** [comment]
I made an empty commit in order to trigger the CI. Hope that can be dropped during PR.

**Review by forivall — APPROVED** [review]

**Comment by forivall** [comment]
@RyanCavanaugh can we get some assistance on the build issues? thanks!

**Comment by RyanCavanaugh** [comment]
Weird, I restarted the build and it failed again. Will look into the logs when I have a chance...

**Review by rbuckton — APPROVED** [review]

**Pull request merged**